File: lib-connector/src/main/java/com/uis/connector/comm/TypeConvert.java
package com.uis.connector.comm;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

/** 获取对象范性Type
 * @author uis
 */
public class TypeConvert {

    public static Type convert(Object target){
        Type[] result = convertAll(target);
        return result != null ? result[0] : null;
    }

    public static Type[] convertAll(Object target){
        Type result[] = null;
        if(target != null) {
            Class cls = target.getClass();
            Type[] types = cls.getGenericInterfaces();
            if (types.length > 0) {
                result = convertType(types[0]);
            }
            if (result == null) {
                result = convertType(cls.getGenericSuperclass());
            }
        }
        return result;
    }

    private static Type[] convertType(Type type) {
        Type[] result = null;
        if (type instanceof ParameterizedType) {
            ParameterizedType paramType = (ParameterizedType) type;
            Type[] params = paramType.getActualTypeArguments();
            List<Type> list = new ArrayList<>();
            if (params.length > 0) {
                int index = 0;
                for(Type t : params) {
                    list.add(t);
                    ++index;
                }
                if(index > 0) {
                    result = new Type[index];
                    list.toArray(result);
                }
            }
        }
        return result;
    }
}

File: assignor/src/main/java/com/uis/assignor/utils/TypeConvert.kt
/*
 * Copyright (c) 2019 by uis
 * Author: uis
 * Github: https://github.com/luiing
 */

package com.uis.assignor.utils

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

/** 获取对象范性Type
 * @author uis
 */
object TypeConvert {

    @JvmStatic
    fun convert(target: Any): Type? = convertAll(target)?.let { it[0] }

    @JvmStatic
    fun convertAll(target: Any): Array<Type>? {
        var result: Array<Type>? = null
        val cls = target.javaClass
        val types = cls.genericInterfaces
        if (types.isNotEmpty()) {
            result = convertType(types[0])
        }
        return result ?: cls.genericSuperclass?.let {
            convertType(it)
        }
    }

    @JvmStatic
    private fun convertType(type: Type): Array<Type>? {
        var result: Array<Type>? = null
        if (type is ParameterizedType) {
            result = type.actualTypeArguments
        }
        return result
    }
}
